tinytodo/TUTORIAL_TEMPLATE.md

```diff
@@ -0,0 +1,232 @@
+
+We'll finish out this tutorial by looking at the changes to the TinyTodo code needed to leverage templates.
+
+### Implementing `List` sharing
+
+Using the Python client, the running example we've been using so far is expressed by the command `share_list(0,interns,read_only=True)` (which follows other commands that created the list 0; see the original tutorial for more). This command will send an HTTP request to the TinyTodo server which induces the `add_share()` function in `context.rs` to be called. Here is its code; it's a lot to digest, so we'll work through it bit by bit.
+
+```rust
+fn add_share(&mut self, r: AddShare) -> Result<AppResponse> {
+    self.is_authorized(&r.uid, &*ACTION_EDIT_SHARE, &r.list)?;
+    // Confirm that the identified list and sharer are known
+    let _list = self.entities.get_list(&r.list)?;
+    let _target_entity = self.entities.get_user_or_team_mut(&r.share_with)?;
+    // Link a template to register the new permission
+    let tid = match r.role {
+        ShareRole::Reader => PolicyId::from_str("reader-template")?,
+        ShareRole::Editor => PolicyId::from_str("editor-template")?,
+    };
+    // Construct template linking environment
+    let target_euid: &cedar_policy::EntityUid = r.share_with.as_ref();
+    let list_euid: &cedar_policy::EntityUid = r.list.as_ref();
+    let env: HashMap<SlotId, cedar_policy::EntityUid> = [
+        (SlotId::principal(), target_euid.clone()),
+        (SlotId::resource(), list_euid.clone()),
+    ]
+    .into_iter()
+    .collect();
+    // Link it!
+    let pid = Self::linked_policy_id(r.role, r.share_with, r.list)?;
+    self.policies.link(tid, pid.clone(), env)?;
+    info!("Created policy {pid}");
+}
+```
+The first line tries to authorize the request; for our example, it checks that `User::"andrew"` can perform `Action::"EditShare"` on resource `List::"0"`. The `is_authorized` call returns `Ok()` because *policy 1* evaluates to `true`, since `List::"0".owner == User::"andrew"` (see Figure 3 in the original tutorial). The call to `get_list` then retrieves the native `List` object for `List::"0"` from `self.entities`, our `EntityStore`. The next call `get_user_or_team_mut` gets the `User` or `Team` to share the list with, which in this case is `Team::"interns"`. If either call fails then the target objects don't exist in the entity store and an error is sent back.
+
+Next, we construct the name of the template we will link: if `r.role` is `Reader`, as in our example, the template ID is `reader-template`. Now we construct the entity UIDs to link `?principal` and `?resource` against, and store them in `env`, a hashmap. Then we call the function `linked_policy_id` (not shown) to construct the policy ID from the sharing role, the target user/team, and the list. This function is injective, so we can be sure no two linked policies will have the same ID. In the case of our example, a policy ID is `reader[interns][0]`. Finally, we link the policy into the policy store.
```

Comment:
> This function is injective, so we can be sure no two linked policies will have the same ID. 

This feels like a bit of a flaw in our template linking API. This shouldn't be the callers responsibility.

Reply:
I agree that it should not be a requirement, but it currently is.

As a concrete suggestion: How about adding an API that constructs the instance policy ID based on the template's ID and the IDs of its linked-in parameters, perhaps with some additional string prefix provided by the caller? E.g.,
```
templateID = reader
?principal = User::"Alice"
?resource = Photo::"vacation.jpg"
prefix = photo
```
yields linked policy ID
```
photo|reader|User::"Alice"|Photo::"vacation.jpg"
```
?

Reply:
After writing this up as an issue, I'm not sure if it's better than the status quo: https://github.com/cedar-policy/cedar/issues/547